### PR TITLE
Add download artifacts in engage publication + add direct links to embed interface 

### PR DESCRIPTION
**Pull request by jduehring**
This PR adds the possibility to add information about the download artifacts to the engage publication channel and display the included direct download links in the admin-ui. 

The publish-engage WOH now has an additional configuration key `list-elements` which defaults to `true` and adds the additional information. If no additional information is preferred, the key must explicitly be set to `false`.

By using the added information, I extended the existing interface for getting embed links and included an option to copy every existing direct link. The extended interface is shown in the screenshot below: 

Screenshot from 2021-06-24 01-43-00

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

File: modules/admin-ui-frontend/app/scripts/shared/partials/modals/embedding-code.html [hunk 1]
```diff
@@ -1,57 +1,112 @@
...
+            success: function (result) {
+              var pubs = result["publications"];
+              for (var i = 0; i < pubs.length; i++) {
+                if (pubs[i]["channel"] == "engage-player") {
+                  var mediaCnt = pubs[i]["media"].length;
+                  var media = pubs[i]["media"];
+                  media.sort((a, b) => (a["width"] > b["width"]) ? -1 : ((b["width"] > a["width"]) ? 1 : 0));
+
+                  //add all necessary divs
+                  for (var j = 0; j < mediaCnt; j++) {
```
Comment (lkiesow) [inline modules/admin-ui-frontend/app/scripts/shared/partials/modals/embedding-code.html:37]: This is simpler and better to read:

```suggestion
                  var media = pubs[i]["media"];
                  media.sort((a, b) => b.width - a.width);

                  // add all necessary divs
                  for (var j = 0; j < media.length; j++) {
```

File: modules/admin-ui-frontend/app/scripts/shared/partials/modals/embedding-code.html [hunk 2]
```diff
@@ -1,57 +1,112 @@
...
+                    downloadLinks[size] = media[j]["url"];
+                    var newBox = document.createElement("div");
+                    switch (j) {
+                      case 0: newBox.classList.add("size_620x349"); break;
+                      case 1: newBox.classList.add("size_540x304"); break;
+                      case 2: newBox.classList.add("size_460x259"); break;
+                      case 3: newBox.classList.add("size_380x214"); break;
+                      case 4: newBox.classList.add("size_300x169"); break;
+                      default: newBox.classList.add("size_300x169");
+                    }
```
Comment (lkiesow) [inline modules/admin-ui-frontend/app/scripts/shared/partials/modals/embedding-code.html:48]: Hard coding this does not work.
We can have arbitrary many resolutions and they can have whatever resolution you like.

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 3]
```diff
@@ -326,6 +330,20 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
       if (!waitForStatus(jobs.toArray(new Job[jobs.size()])).isSuccess()) {
         throw new WorkflowOperationException("One of the distribution jobs did not complete successfully");
       }
+      //Prepare download artifacts to be added
+      Set<MediaPackageElement> mediaPackageData = new HashSet<>();
+      if (addData) {
+        for (Job job : jobs) {
+          try {
+            List<? extends MediaPackageElement> elems = MediaPackageElementParser.getArrayFromXml(job.getPayload());
+            mediaPackageData.addAll(elems);
```
Comment (lkiesow) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:339]:
```suggestion
            mediaPackageData.addAll(MediaPackageElementParser.getArrayFromXml(job.getPayload()));
```

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 4]
```diff
@@ -213,6 +214,9 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
     boolean checkAvailability = option(op.getConfiguration(CHECK_AVAILABILITY)).bind(trimToNone).map(toBool)
             .getOrElse(true);
 
+    boolean addData = option(op.getConfiguration(LIST_ELEMENTS)).bind(trimToNone).map(toBool)
```
Comment (lkiesow) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:217]: The variable name is pretty much meaningless. What about using:

```suggestion
    boolean publishElements = option(op.getConfiguration(LIST_ELEMENTS)).bind(trimToNone).map(toBool)
```

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 5]
```diff
@@ -326,6 +330,20 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
       if (!waitForStatus(jobs.toArray(new Job[jobs.size()])).isSuccess()) {
         throw new WorkflowOperationException("One of the distribution jobs did not complete successfully");
       }
+      //Prepare download artifacts to be added
+      Set<MediaPackageElement> mediaPackageData = new HashSet<>();
```
Comment (lkiesow) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:334]:
```suggestion
      Set<MediaPackageElement> mediaPackageElements = new HashSet<>();
```

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 6]
```diff
@@ -326,6 +330,20 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
...
+      if (addData) {
+        for (Job job : jobs) {
+          try {
+            List<? extends MediaPackageElement> elems = MediaPackageElementParser.getArrayFromXml(job.getPayload());
+            mediaPackageData.addAll(elems);
+          } catch (MediaPackageException e) {
+            logger.error("Job '{}' returned payload ({}) that could not be parsed to media package elements", job,
+                job.getPayload(), e);
+            throw new WorkflowOperationException(e);
```
Comment (lkiesow) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:343]:
```suggestion
            throw new WorkflowOperationException(String.format("Job '%s' returned payload (%s) that could not be parsed to media package elements", job,
                job.getPayload()), e);
```

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 7]
```diff
@@ -375,6 +393,13 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
         // Create new distribution element
         Publication publicationElement = PublicationImpl.publication(UUID.randomUUID().toString(), CHANNEL_ID,
                 engageUri, MimeTypes.parseMimeType("text/html"));
+
+        //Add download Artifacts
+        for (MediaPackageElement element : mediaPackageData) {
+          element.setIdentifier(null);
+          logger.info("HERE URI " + element.getURI());
```
Comment (lkiesow) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:400]:
```suggestion
```

**Review by lkiesow — CHANGES_REQUESTED** [review]
The backend looks good. My comments are just nitpicks.
Can you maybe publish it as a separate pull request, so we can get that already in?

The frontend has a few problems. But one additional problem I see it that the existing code should never have made it into Opencast like this. It should have been a separate JavaScript file with AngularJS code instead. Let's talk about what we can do about that…

File: modules/admin-ui-frontend/app/scripts/shared/partials/modals/embedding-code.html (hunk 2 above)
Comment (jduehring) [reply modules/admin-ui-frontend/app/scripts/shared/partials/modals/embedding-code.html:48]: I believe this shouldn't be an issue, since this code only sets the size of the clickable boxes. The ratio of those boxes wasn't exactly tied to their real size, so I thought I could reuse them to just show different sizes at all. 

Should the ratio of the displayed boxes exactly match the ratio of the actual resolutions?

File: docs/guides/admin/docs/workflowoperationhandlers/publish-engage-woh.md [hunk 8]
```diff
@@ -30,6 +30,7 @@ Parameter Table
 |streaming-target-subflavors|Subflavor to use for distributed material                                                    |
 |merge-force-flavors        |Flavors of elements for which an update is enforced when merging catalogs.                  |
 |                           |Defaults to `dublincore/*,security/*`.
+|list-elements              |Include all media package elements in the publication. Defaults to `true`.
```
Comment (mtneug) [inline docs/guides/admin/docs/workflowoperationhandlers/publish-engage-woh.md:33]: Does it make sense to deactivate this? There is a | missing at the end.

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java (hunk 4 above)
Comment (mtneug) [reply modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:217]: `publich -> publish`

**Review by mtneug — CHANGES_REQUESTED** [review]
How does this interact with the `retract-partial` WOH?

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java (hunk 4 above)
Comment (mtneug) [reply modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:217]: `publich -> publish`

**Comment by lkiesow** [comment]
> How does this interact with the retract-partial WOH?

Good point. That one should remove tracks it removes.
That should be in:

    modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PartialRetractEngageWorkflowOperationHandler.java

File: docs/guides/admin/docs/workflowoperationhandlers/publish-engage-woh.md (hunk 8 above)
Comment (mtneug) [reply docs/guides/admin/docs/workflowoperationhandlers/publish-engage-woh.md:33]: The consensus during the dev meeting seems to be to always enable listing the published elements.

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 9]
```diff
@@ -375,6 +393,13 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
         // Create new distribution element
         Publication publicationElement = PublicationImpl.publication(UUID.randomUUID().toString(), CHANNEL_ID,
                 engageUri, MimeTypes.parseMimeType("text/html"));
+
+        //Add download Artifacts
```
Comment (mtneug) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:397]:
```suggestion
        // Add published elements
```

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java [hunk 10]
```diff
@@ -326,6 +330,20 @@ public WorkflowOperationResult start(final WorkflowInstance workflowInstance, Jo
       if (!waitForStatus(jobs.toArray(new Job[jobs.size()])).isSuccess()) {
         throw new WorkflowOperationException("One of the distribution jobs did not complete successfully");
       }
+      //Prepare download artifacts to be added
```
Comment (mtneug) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/PublishEngageWorkflowOperationHandler.java:333]:
```suggestion
      // Prepare published elements to be added
```

**Review by mtneug — CHANGES_REQUESTED** [review]

File: modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/ConfigurablePublishWorkflowOperationHandler.java [hunk 11]
```diff
@@ -346,7 +347,7 @@ private boolean distributeElements(SimpleElementSelector selector, MediaPackage
     return true;
   }
 
-  private Set<MediaPackageElement> distribute(
+  protected Set<MediaPackageElement> distribute(
```
Comment (mtneug) [inline modules/distribution-workflowoperation/src/main/java/org/opencastproject/workflow/handler/distribution/ConfigurablePublishWorkflowOperationHandler.java:350]: Why is this necessary?

**Review by mtneug — CHANGES_REQUESTED** [review]

**Comment by mtneug** [comment]
Can we close this in favour of #2811?

**Comment by gregorydlogan** [comment]
This adds the UI elements, does it not?  IMO this is likely an extension of #2811, and might be worth a rebase.

**Comment by gregorydlogan** [comment]
Per discussion in the technical meeting, we're going to close this in favour of the split PRs.
